src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerGenotypingEngine.java

```diff
@@ -629,39 +626,36 @@ protected static void cleanUpSymbolicUnassembledEvents( final List<Haplotype> ha
-        }
+    protected static Map<Allele, List<Haplotype>> createAlleleMapper(final List<VariantContext> eventsAtThisLoc,
+                                                                     final VariantContext mergedVC,
+                                                                     final int loc,
+                                                                     final List<Haplotype> haplotypes) {
+        Utils.validateArg(haplotypes.size() > eventsAtThisLoc.size(), "expected haplotypes.size() >= eventsAtThisLoc.size() + 1");
 
-        return mergeMap.entrySet().stream().collect(Collectors.toMap(e -> e.getValue(), e -> eventMapper.get(new Event(e.getKey())), (x,y)->x, LinkedHashMap<Allele, List<Haplotype>>::new));
-    }
+        final Map<Allele, List<Haplotype>> result = new LinkedHashMap<>();
 
-    protected static Map<Event, List<Haplotype>> createEventMapper(final int loc, final List<VariantContext> eventsAtThisLoc, final List<Haplotype> haplotypes) {
-        Utils.validateArg(haplotypes.size() > eventsAtThisLoc.size(), "expected haplotypes.size() >= eventsAtThisLoc.size() + 1");
-        final Map<Event, List<Haplotype>> eventMapper = new LinkedHashMap<>(eventsAtThisLoc.size()+1);
-        final Event refEvent = new Event(null);
-        eventMapper.put(refEvent, new ArrayList<>());
-        for( final VariantContext vc : eventsAtThisLoc ) {
-            eventMapper.put(new Event(vc), new ArrayList<>());
-        }
+        final Allele ref = mergedVC.getReference();
+        result.put(ref, new ArrayList<>());
 
-        for( final Haplotype h : haplotypes ) {
-            if( h.getEventMap().get(loc) == null ) {
-                eventMapper.get(refEvent).add(h);
+        //Note: we can't use the alleles implied by eventsAtThisLoc because they are not yet merged to a common reference
+        //For example, a homopolymer AAAAA reference with a single and double deletion would yield (i) AA* A and (ii) AAA* A
+        //in eventsAtThisLoc, when in mergedVC it would yield AA* AA A
+        mergedVC.getAlternateAlleles().stream().filter(a -> !a.isSymbolic()).forEach(a -> result.put(a, new ArrayList<>()));
+
+        for (final Haplotype h : haplotypes) {
+            final VariantContext haplotypeVC = h.getEventMap().get(loc);
+            if (haplotypeVC == null) {    //no events --> this haplotype supports the reference at this locus
+                result.get(ref).add(h);
             } else {
-                for( final VariantContext vcAtThisLoc : eventsAtThisLoc ) {
-                    if( h.getEventMap().get(loc).hasSameAllelesAs(vcAtThisLoc) ) {
-                        eventMapper.get(new Event(vcAtThisLoc)).add(h);
+                for (int n = 0; n < eventsAtThisLoc.size(); n++) {
+                    if (haplotypeVC.hasSameAllelesAs(eventsAtThisLoc.get(n))) {
```

Comment:
Notes:

- `eventsAtThisLoc` and `haplotypes` share the same variant contexts (alleles). mergedVC may have alleles that those two don't have (see David's example above). 
- mixing the index `n` between `mergedVC` and `eventsAtThisLoc` may be dangerous as hinted by the comment in original code